### PR TITLE
enhance: use unique_ptr in Config

### PR DESCRIPTION
**Pull request by jackwener**
Part of #663 

In config.cc, use unique_ptr to eliminate some trivial manual deallocation

I have remove the vector [List-initialization](https://en.cppreference.com/w/cpp/language/list_initialization).

I use `[]` place it

### CODE REVIEW

**Comment by PragmaTwice** [comment]
```c++
template<typename T, typename ...Args> auto UniqueConfig(Args &&... args) {
  return std::unique_ptr<ConfigField>(new T(std::forward<Args>(args)...));
}
```
You can add this utility function to smplify the changes : )

**Comment by jackwener** [comment]
> ```c++
> template<typename T, typename ...Args> auto UniqueConfig(Args &&... args) {
>   return std::unique_ptr<ConfigField>(new T(std::forward<Args>(args)...));
> }
> ```
> 
> You can add this utility function to smplify the changes : )

Looks like changing a constructor is clearest. I have done it.

File: src/config.cc [hunk 1]
```diff
@@ -84,11 +84,15 @@ const char *configEnumGetName(configEnum *ce, int val) {
...
   struct FieldWrapper {
-    const char *name;
+    const std::string name;
     bool readonly;
-    ConfigField *field;
+    std::unique_ptr<ConfigField> field;
+
+    FieldWrapper(std::string name, bool readonly,
+                 ConfigField* field)
+        : name(std::move(name)), readonly(readonly), field(std::move(field)) {}
```
Comment (PragmaTwice) [inline src/config.cc]: Applying `std::move` to a pointer `ConfigField* field` is useless.

File: src/config.cc [hunk 2]
```diff
@@ -484,9 +488,6 @@ void Config::initFieldCallback() {
 }
 
 Config::~Config() {
```
Comment (PragmaTwice) [inline src/config.cc]: Could you delete it and add `~Config() = default;` in `config.h`?
We'd better make sure these trivial function in header to let compilers do more optimization like inlining and call elimination.

**Comment by jackwener** [comment]
Thanks @PragmaTwice ! ❤️👍

File: src/config.cc [hunk 3]
```diff
@@ -181,10 +185,10 @@ Config::Config() {
       {"rocksdb.level_compaction_dynamic_level_bytes",
         false, new YesNoField(&RocksDB.level_compaction_dynamic_level_bytes, false)},
   };
-  for (const auto &wrapper : fields) {
-    auto field = wrapper.field;
+  for (auto &wrapper : fields) {
+    auto &field = wrapper.field;
     field->readonly = wrapper.readonly;
-    fields_.insert({wrapper.name, field});
+    fields_.insert({wrapper.name, std::move(field)});
```
Comment (PragmaTwice) [inline src/config.cc]:
```suggestion
    fields_.emplace(std::move(wrapper.name), std::move(field));
```
We can also move the name string 🤣

**Review by PragmaTwice — DISMISSED** [review]
LGTM

File: src/config.cc [hunk 4]
```diff
@@ -84,11 +84,15 @@ const char *configEnumGetName(configEnum *ce, int val) {
 
 Config::Config() {
   struct FieldWrapper {
-    const char *name;
+    const std::string name;
```
Comment (PragmaTwice) [inline src/config.cc]:
```suggestion
    std::string name;
```

Comment (tisonkun) [reply src/config.cc]: Why do you make this change?

**Review by PragmaTwice — APPROVED** [review]
LGTM

**Review by git-hulk — APPROVED** [review]

File: src/config.cc (hunk 4 above)
Comment (PragmaTwice) [reply src/config.cc]: `const std::string name` may prevent `std::move` from FieldWrapper::name to key of `std::map field_` in https://github.com/apache/incubator-kvrocks/blob/8836bc57d18927534c928d0b8b7d19a5d93c9d5f/src/config.cc#L191

**Review by tisonkun — APPROVED** [review]

**Comment by PragmaTwice** [comment]
Thanks for you contribution! Merging...

**Pull request merged**